solutions/usaco-646.mdx

```diff
@@ -26,7 +26,96 @@ Note: The video solution might not be the same as other solutions. Code in C++.
 **Time Complexity:** $\mathcal{O}(V \cdot E \cdot \alpha(n)))$
 
 <LanguageSection>
+<CPPSection>
 
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+
+//BeginCodeSnip{DSU}
+struct DSU {
+	vector<int> e;
+	DSU(int N) { e = vector<int>(N, -1); }
+
+	// get representive component (uses path compression)
+	int get(int x) { return e[x] < 0 ? x : e[x] = get(e[x]); }
+
+	bool same_set(int a, int b) { return get(a) == get(b); }
+
+	int size(int x) { return -e[get(x)]; }
+
+	bool unite(int x, int y) {	// union by size
+		x = get(x), y = get(y);
+		if (x == y) return false;
+		if (e[x] > e[y]) swap(x, y);
+		e[x] += e[y];
+		e[y] = x;
+		return true;
+	}
+};
+//EndCodeSnip
+
+int main() {
+	freopen("closing.in", "r", stdin);
+	freopen("closing.out", "w", stdout);
+
+	int n, m;
+	cin >> n >> m;
+
+	vector<vector<int>> adj(n);
+	for (int i = 0; i < m; i++) {
+		int u, v;
+		cin >> u >> v;
+		--u;
+		--v;
+		adj[u].push_back(v);
+		adj[v].push_back(u);
+	}
+
+	// conn[i] = whether the ith farm is closed
+	vector<bool> conn(n);
+	vector<int> rev(n);
+	for (int i = 0; i < n; i++) {
+		cin >> rev[i];
+		rev[i]--;
+	}
+
+	DSU dsu(n);
+	reverse(rev.begin(), rev.end());
+
+	conn[rev[0]] = 1;
+	// one node is always connected
+	vector<string> ans = {"YES"};
+
+	// connected components
+	int cc = 1;
+	for (int i = 1; i < n; i++) {
+		cc++;
+		conn[rev[i]] = 1;
+		for (int j : adj[rev[i]]) {
+			if (conn[j]) {
+				if (dsu.unite(j, rev[i])) {
+					cc--;
+				}
+			}
+		}
+
+		if (cc == 1) {
+			ans.push_back("YES");
+		} else {
+			ans.push_back("NO");
+		}
+	}
+	reverse(ans.begin(), ans.end());
+
+	for (auto i : ans) {
+		cout << i << endl;
```

Comment:
i know i said i like endl but in a for loop i've decided that \n is better

Reply:
burh